models/aeif_cbvg_2010.cpp

```diff
@@ -0,0 +1,657 @@
+
+/* ----------------------------------------------------------------
+ * Recordables map
+ * ---------------------------------------------------------------- */
+
+nest::RecordablesMap< nest::aeif_cbvg_2010 >
+  nest::aeif_cbvg_2010::recordablesMap_;
+
+namespace nest
+{
+/*
+ * template specialization must be placed in namespace
+ *
+ * Override the create() method with one call to RecordablesMap::insert_()
+ * for each quantity to be recorded.
+ */
+template <>
+void
+RecordablesMap< aeif_cbvg_2010 >::create()
+{
+  // use standard names whereever you can for consistency!
+  insert_(
+    names::V_m, &aeif_cbvg_2010::get_y_elem_< aeif_cbvg_2010::State_::V_M > );
+  insert_(
+    names::w, &aeif_cbvg_2010::get_y_elem_< aeif_cbvg_2010::State_::W > );
+  insert_(
+    names::z, &aeif_cbvg_2010::get_y_elem_< aeif_cbvg_2010::State_::Z > );
+  insert_(
+    names::V_T, &aeif_cbvg_2010::get_y_elem_< aeif_cbvg_2010::State_::V_T > );
+  insert_( names::u_bar_plus,
+    &aeif_cbvg_2010::get_y_elem_< aeif_cbvg_2010::State_::U_BAR_PLUS > );
+  insert_( names::u_bar_minus,
+    &aeif_cbvg_2010::get_y_elem_< aeif_cbvg_2010::State_::U_BAR_MINUS > );
+  insert_( names::u_bar_bar,
+    &aeif_cbvg_2010::get_y_elem_< aeif_cbvg_2010::State_::U_BAR_BAR > );
+}
+}
+
+
+extern "C" int
+nest::aeif_cbvg_2010_dynamics( double,
+  const double y[],
+  double f[],
+  void* pnode )
+{
+  // a shorthand
+  typedef nest::aeif_cbvg_2010::State_ S;
+
+  // get access to node so we can almost work as in a member function
+  assert( pnode );
+  const nest::aeif_cbvg_2010& node =
+    *( reinterpret_cast< nest::aeif_cbvg_2010* >( pnode ) );
+
+  const bool is_refractory = node.S_.r_ > 0;
+  const bool is_clamped = node.S_.clamp_r_ > 0;
+
+  // y[] here is---and must be---the state vector supplied by the integrator,
+  // not the state vector in the node, node.S_.y[].
+
+  // The following code is verbose for the sake of clarity. We assume that a
+  // good compiler will optimize the verbosity away...
+
+  // Clamp membrane potential to V_reset while refractory, otherwise bound
+  // it to V_peak.
+  const double& V = ( is_refractory || is_clamped )
+    ? ( is_clamped ? node.P_.V_clamp_ : node.P_.V_reset_ )
+    : std::min( y[ S::V_M ], node.P_.V_peak_ );
+  // shorthand for the other state variables
+  const double& w = y[ S::W ];
+  const double& z = y[ S::Z ];
+  const double& V_T = y[ S::V_T ];
+  const double& u_bar_plus = y[ S::U_BAR_PLUS ];
+  const double& u_bar_minus = y[ S::U_BAR_MINUS ];
+  const double& u_bar_bar = y[ S::U_BAR_BAR ];
+
+  const double I_spike = ( node.P_.g_L * node.P_.Delta_T
+    * std::exp( ( V - V_T ) / node.P_.Delta_T ) );
+
+  // dv/dt
+  f[ S::V_M ] = ( is_refractory || is_clamped )
+    ? 0.0
+    : ( -node.P_.g_L * ( V - node.P_.E_L ) + I_spike - w + z + node.P_.I_e
+        + node.B_.I_stim_ ) / node.P_.C_m;
+
+  // Adaptation current w.
+  f[ S::W ] =
+    is_clamped ? 0.0 : ( node.P_.a * ( V - node.P_.E_L ) - w ) / node.P_.tau_w;
```

Comment:
Why would ``w`` be fixed when the neuron is clamped? I don't see how you would fix both current and voltage...

Reply:
@Silmathoron We acuatlly adapted this from the original implementation by Claudia Clopath. I am not sure how to proceed here. Do you think this needs to be changed?

Reply:
Hey, sorry for the long delay, I read the notification at a time I could not reply, then forgot about it...
Skimming briefly through the 2 papers, I did not see where this clamping of `w` occurs, could you tell me precisely where it is detailed?

Reply:
Unfortunately, the whole clamping procedure is not described in the paper. We adapted it from the code that we found on [ModelDB](https://senselab.med.yale.edu/modeldb/showModel.cshtml?model=144566&file=%2fmodeldb_package%2fVoTriCode%2faEIF.m). There, the temporal evolution of `w` during clamping is undone in the line `w = w-wdot;`.

Reply:
Ok, well, I guess if this is part of the implementation on ModelDB then the NEST implementation should also conform to it...
Could you also link to ModelDB in the docstring, please? I think this would be quite useful

Reply:
Agreed! We added the references with 4801307.